File: frontend/src/api/endpoints/stats.js
import { db, requireAdmin } from "../db.js"
import { ApiError, ErrorCode } from "../errors.js"
import { STATUS_LABELS, Status } from "../types.js"
import { voteRows } from "./requests.js"

function pad(n) {
  return String(n).padStart(2, "0")
}

function toDayKey(iso) {
  const d = new Date(iso)
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`
}

export function getStats() {
  const currentUser = requireAdmin()
  if (!currentUser) throw new ApiError("Admin access required", ErrorCode.FORBIDDEN)

  const totalRequests = db.requests.length
  const activeRequests = db.requests.filter((r) =>
    [Status.UNDER_REVIEW, Status.PLANNED, Status.IN_PROGRESS].includes(r.status)
  ).length
  const totalSupport = db.votes.length
  const releasedRequests = db.requests.filter((r) => r.status === Status.RELEASED).length

  const statusDistribution = Object.values(Status).map((status) => ({
    status,
    label: STATUS_LABELS[status],
    count: db.requests.filter((r) => r.status === status).length,
  }))

  const topVoted = [...db.requests]
    .sort((a, b) => voteRows(b.id).length - voteRows(a.id).length || (a.createdAt < b.createdAt ? 1 : -1))
    .slice(0, 5)
    .map((r) => ({
      id: r.id,
      title: r.title,
      status: r.status,
      support: voteRows(r.id).length,
    }))

  // Votes per day over the last 30 days.
  const now = new Date()
  const buckets = new Map()
  for (let i = 29; i >= 0; i -= 1) {
    const d = new Date(now)
    d.setDate(d.getDate() - i)
    buckets.set(toDayKey(d.toISOString()), 0)
  }
  for (const vote of db.votes) {
    const key = toDayKey(vote.createdAt)
    if (buckets.has(key)) buckets.set(key, buckets.get(key) + 1)
  }
  const votesOverTime = [...buckets.entries()].map(([date, count]) => ({ date, count }))

  return {
    totalRequests,
    activeRequests,
    totalSupport,
    releasedRequests,
    statusDistribution,
    topVoted,
    votesOverTime,
  }
}

File: backend/routers/stats.py
from datetime import datetime, timedelta, timezone

from fastapi import APIRouter, Depends
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from auth import require_admin
from database import get_db
from models.request import Request
from models.user import User
from models.vote import Vote
from schemas import Stats, StatusBucket, TopVoted, VotesOverTime
from schemas.common import ACTIVE_STATUSES, STATUS_LABELS, STATUS_RELEASED
from serializers import support_count

router = APIRouter(tags=["stats"])


@router.get("/stats", response_model=Stats)
def get_stats(
    admin: User = Depends(require_admin),
    db: Session = Depends(get_db),
) -> Stats:
    total_requests = db.scalar(select(func.count()).select_from(Request)) or 0
    active_requests = (
        db.scalar(
            select(func.count())
            .select_from(Request)
            .where(Request.status.in_(ACTIVE_STATUSES))
        )
        or 0
    )
    total_support = db.scalar(select(func.count()).select_from(Vote)) or 0
    released_requests = (
        db.scalar(
            select(func.count())
            .select_from(Request)
            .where(Request.status == STATUS_RELEASED)
        )
        or 0
    )

    status_distribution = []
    for status, label in STATUS_LABELS.items():
        count = (
            db.scalar(
                select(func.count())
                .select_from(Request)
                .where(Request.status == status)
            )
            or 0
        )
        status_distribution.append(
            StatusBucket(status=status, label=label, count=count)
        )

    requests = db.scalars(select(Request)).all()
    supports = {r.id: support_count(db, r.id) for r in requests}
    top = sorted(
        requests, key=lambda r: (supports[r.id], r.created_at), reverse=True
    )[:5]
    top_voted = [
        TopVoted(id=r.id, title=r.title, status=r.status, support=supports[r.id])
        for r in top
    ]

    today = datetime.now(timezone.utc).date()
    buckets = {(today - timedelta(days=i)).isoformat(): 0 for i in range(29, -1, -1)}
    for created_at in db.scalars(select(Vote.created_at)).all():
        key = created_at[:10]
        if key in buckets:
            buckets[key] += 1
    votes_over_time = [
        VotesOverTime(date=date, count=count) for date, count in buckets.items()
    ]

    return Stats(
        total_requests=total_requests,
        active_requests=active_requests,
        total_support=total_support,
        released_requests=released_requests,
        status_distribution=status_distribution,
        top_voted=top_voted,
        votes_over_time=votes_over_time,
    )